docs: transfer set up from a covenant signer to a covenant emulator (#50)
babylonlabs-io/covenant-emulator: covenant-signer/README.md

> # Covenant Signer

Reviewer: Phase-1 covenants have also used the Phase-1 "Covenant Signer" (github.com/babylonlabs-io/covenant-signer), which is a completely different program than this. To avoid confusion, maybe we should add some small letters note here that this is a different program.

## Table of Contents

1. [Prerequisites](#1-prerequisites)
2. [Install covenant signer](#2-install-covenant-signer-binary)
3. [Export the key from the Bitcoin node](#3-export-the-key-from-the-bitcoin-node)
4. [Import the key into the cosmos keyring](#4-import-the-key-into-the-cosmos-keyring)
5. [Create the configuration file](#5-create-the-configuration-file)
6. [Running the Covenant Signer](#6-running-the-covenant-signer)
7. [Using the covenant signer for signing transactions](#7-using-the-covenant-signer-for-signing-transactions)

## 1. Prerequisites

> To successfully complete this guide, ensure you have the bitcoin node setup and you 
> have access to the private Bitcoin key you used previously when you generated your 
> address as it contains the covenant key.

Reviewer: ```suggestion
This guide requires that
1. you have a Bitcoin node setup for the Bitcoin
network you intend to operate your covenant signer in and
2. you have access to the the private Bitcoin key you
set up your covenant with.
```

For a refresher on setting up the bitcoin node, refer to the 
[deployment guide](https://github.com/babylonlabs-io/covenant-signer/blob/main/docs/deployment.md#2-bitcoind-setup).

## 2. Install covenant signer binary

> If you havent already, download [Golang 1.21](https://go.dev/dl) 

Reviewer: We have golang 1.23 in this repo

  "pubkey": "023a79b546c79d7f7c5ff20620d914b5cf7250631d12f6e26427ed9d3f98c5ccb1",
  "ischange": false,
  "timestamp": 1732624709,
  "hdkeypath": "m/84h/1h/0h/0/0",
  "hdseedid": "0000000000000000000000000000000000000000",
  "hdmasterfingerprint": "5e174bde",
  "labels": [
    ""
  ]
}
```

The most important field to focus on is `hdkeypath` that contains derivation path 
of our key. In the example it is `84h/1h/0h/0/0` (the intilal `m/` can be ignored).

Next, list all descriptors in the wallet, ensuring that private keys are included 
in the output:

```shell
docker exec -it bitcoind bitcoin-cli -chain=regtest -rpcuser=user -rpcpassword=pass listdescriptors true
```

The terminal should produce output similar to the following:

```json
{
  "wallet_name": "covenant-wallet",
  "descriptors": [
    {
      "desc": "wpkh(tprv8ZgxMBicQKsPe9aCeUQgMEMy2YMZ6PHnn2iCuG12y5E8oYhYNEvUqUkNy6sJ7ViBmFUMicikHSK2LBUNPx5do5EDJBjG7puwd6azci2wEdq/84h/1h/0h/0/*)#sachkrde",
      "timestamp": 1732624709,
      "active": true,
      "internal": false,
      "range": [
        0,
        1000
      ],
      "next": 1,
      "next_index": 1
    }
    ...
  ]
}

> 

Reviewer: ```suggestion
```

```

The most important field to note is the `desc` value:

```json
"desc": "wpkh(tprv8ZgxMBicQKsPe9aCeUQgMEMy2YMZ6PHnn2iCuG12y5E8oYhYNEvUqUkNy6sJ7ViBmFUMicikHSK2LBUNPx5do5EDJBjG7puwd6azci2wEdq/84h/1h/0h/0/*)#sachkrde"
```

> Here, you can see the string starting with `tprv8ZgxMBicQKsPe9aCeUQgMEMy2YMZ6PHnn2iCuG12y5E8oYhYNEvUqUkNy6sJ7ViBmFUMicikHSK2LBUNPx5do5EDJBjG7puwd6azci2wEdq`is the **base58-encoded master private key** of the covenant wallet. This key is critical for signing operations and should be securely stored.

Reviewer: Let's wrap all lines to 80 characters

#### Deriving the Covenant Private Key from the Master Key

You can derive the covenant private key from the master key by performing a **BIP32 derivation**. The `covenant-signer`repository includes a command to accomplish this:

```shell

> covenant-signer derive-child-key \ tprv8ZgxMBicQKsPe9aCeUQgMEMy2YMZ6PHnn2iCuG12y5E8oYhYNEvUqUkNy6sJ7ViBmFUMicikHSK2LBUNPx5do5EDJBjG7puwd6azci2wEdq \ 84h/1h/0h/0/0

Reviewer: formatting

    name: cov
    pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"Ajp5tUbHnX98X/IGINkUtc9yUGMdEvbiZCftnT+Yxcyx"}'
    type: local

```

Here, the `key` field contains the base64-encoded public key. After decoding, 
this key:

```shell
023a79b546c79d7f7c5ff20620d914b5cf7250631d12f6e26427ed9d3f98c5ccb1
```

Matches the public key derived earlier and seen in the outputs of `getaddressinfo` 
and `derive-child-key`.

Congratulations! You have successfully imported your keys from the prior setup 
and verified your setup for the covenant emulator.

## 5. Create the configuration file

Use the example configuration [file](./example/config.toml) to create your own 
configuration file. Then, replace the placeholder values with your own configuration.

```toml
[keystore]
keystore-type = "cosmos"

[keystore.cosmos]
# pointing to the directory where the key is stored, unless specified otherwise
key-directory = "/path/to/keydir"

# the backend to be used for storing the key, in this case file
keyring-backend = "file"

# the name of the key you used when importing the key
key-name = "your-key-name"

# the chain id of the chain to be used
chain-id = "your-chain-id"

> [server-config]

Reviewer: Let's denote what this means?

host = "127.0.0.1"
port = 9791

> [metrics]

Reviewer: Same here

host = "127.0.0.1"
port = 2113
```

> 

Reviewer: We should have some explanations on the key values here.

## 6. Running the Covenant Signer

The covenant signer can be run using the following command:

```shell
covenant-signer start --config ./path/to/config.toml
```

> the covenant signer must be run in a secure network and only accessible by the 

Reviewer: ```suggestion
The covenant signer must be run in a secure network and only accessible by the 
```

covenant emulator.

Once the covenant signer is set up and unlocked, you can configure the covenant 
emulator to use it. The URL of the covenant signer, which is http://127.0.0.1:9791, 
should be specified in the covenant emulator's configuration file under the 
remotesigner section.

It's important to note that the key specified in the covenant emulator's 
configuration is not the covenant key itself. Instead, it is a 
key used for sending Cosmos transactions.

> 

Reviewer: We have never specified what this program is and why it exists. We should have an intro that describes the program as:
- a signer program, intended to be used by members of the covenant committee
- intended to contain only the bare minimum as it has access to the private key -- it should be highly secured
- the application logic for the covenant member is implemented covenant-emulator program that interacts with the Babylon blockchain to receive things to sign and forwards them to the covenant signer to receive sigs.

## 7. Using the covenant signer for signing transactions

> To enable signing transactions with the covenant key, you need to unlock it first.

Reviewer: You need to unlock the keyring it contains. We should denote what the password corresponds to.

```shell
curl -X POST http://127.0.0.1:9791/v1/unlock -d '{"passphrase": "<passphrase>"}'
```

Now that the key is unlocked you can add its url to the covenant emulator's 

> configuration file. See [here](./docs/transfer-setup.md#42-configure-the-covenant-emulator) 

Reviewer: I don't see such docs here

```

Here, you can see the string starting with `tprv8ZgxMBicQKsPe9aCeUQgMEMy2YMZ6PHnn2iCuG12y5E8oYhYNEvUqUkNy6sJ7ViBmFUMicikHSK2LBUNPx5do5EDJBjG7puwd6azci2wEdq`
is the **base58-encoded master private key** of the covenant wallet. 
This key is critical for signing operations and should be securely stored.

#### Deriving the Covenant Private Key from the Master Key

You can derive the covenant private key from the master key by performing a 
**BIP32 derivation**. The `covenant-signer`repository includes a command to 
accomplish this:

```shell
covenant-signer derive-child-key \
  tprv8ZgxMBicQKsPe9aCeUQgMEMy2YMZ6PHnn2iCuG12y5E8oYhYNEvUqUkNy6sJ7ViBmFUMicikHSK2LBUNPx5do5EDJBjG7puwd6azci2wEdq \
  84h/1h/0h/0/0
```

The output will display the derived private and public keys:

```shell
Derived private key: fe1c56c494c730f13739c0655bf06e615409870200047fc65cdf781837cf7f06
Derived public key: 023a79b546c79d7f7c5ff20620d914b5cf7250631d12f6e26427ed9d3f98c5ccb1
```

As seen, the **Derived Public Key**:

```
023a79b546c79d7f7c5ff20620d914b5cf7250631d12f6e26427ed9d3f98c5ccb1
```

Matches the public key obtained earlier using the `getaddressinfo` command.

> ## 4. Importing the Derived Private Key into the Cosmos Keyring

Reviewer 2: I would add here that:
- `file` backend should be used i.e private key will be stored in encrypted form on disk
-  when running `import-hex`  on encrypted file backend, caller will be asked for passphrase which will be later used to unlock the signer

- `keystore-type`: Type of keystore used, e.g., "cosmos" for Cosmos SDK compatibility.
- `key-directory`: Path where keys are stored on the filesystem.
- `keyring-backend`: Backend system for key management, e.g., "file", "os".
- `key-name`: Name of the key used for signing transactions.
- `chain-id`: Identifier of the blockchain network.
- `host` (server-config): IP address where the server listens, typically "127.0.0.1" for local access.
- `port` (server-config): TCP port number for the server.
- `host` (metrics): IP address for the Prometheus metrics server, typically "127.0.0.1".
- `port` (metrics): TCP port number for the Prometheus metrics server.

## 6. Running the Covenant Signer

The covenant signer can be run using the following command:

```shell
covenant-signer start --config ./path/to/config.toml
```

The covenant signer must be run in a secure network and only accessible by the 
covenant emulator.

Once the covenant signer is set up and unlocked, you can configure the covenant 
emulator to use it. The URL of the covenant signer, which is 
http://127.0.0.1:9791, should be specified in the covenant emulator's 
configuration file under the `remotesigner` section.

> It's important to note that the key specified in the covenant emulator's 

Reviewer: Right now we are at the covenant signer, why are we referring to the covenant emulator?

configuration is not the covenant key itself. Instead, it is a 
key used for sending Cosmos transactions.

> ## 7. Using the covenant signer for signing transactions

Reviewer: I think this guide mixes transition and operational concepts together without separating them.

I recommend that this document is split in the following sections:

1. Pre-requisites
2. Installation
3. Transitioning your covenant key from the phase-1 setup
4. Operation
    a. Config
    b. Starting the daemon
    c. Unlocking the key
    d. Testing the setup

Before you can sign transactions with the covenant key, you need to unlock the 
keyring that stores it. The passphrase here must match the one you used when 
importing the covenant key into the keyring. Once the keyring is unlocked, you 
can use the covenant key to sign transactions.

```shell
curl -X POST http://127.0.0.1:9791/v1/unlock -d '{"passphrase": "<passphrase>"}'
```

Now that the key is unlocked you can add its url to the covenant emulator's 
configuration file. See [here](../docs/transfer-setup.md#42-configure-the-covenant-emulator) 
for more information.

You can sign transactions using the following command. However, ensure that both 
the staking and unbonding transactions are provided in hex format.

```shell
curl -X POST http://127.0.0.1:9791/v1/sign-transactions \

>   -d '{

Reviewer 2: as mentioned on slack this json is slightly bigger i.e something like:
```
{
    "staking_tx_hex": "020000000001",
    "slashing_tx_hex": "020000000001...",
    "unbonding_tx_hex": "020000000001...",
    "slash_unbonding_tx_hex": "020000000001...",
    "staking_output_idx": 0,
    "slashing_script_hex": "76a914...",
    "unbonding_script_hex": "76a914...",
    "unbonding_slashing_script_hex": "76a914...",
    "fp_enc_keys": [
        "0123456789abcdef..."
    ]
}
```
